R/cellbrowser.R

```diff
@@ -0,0 +1,499 @@
+# Build a UCSC cell browser website from a \code{Seurat} object
+#
+NULL
+# require(reticulate)
+
+#' Used by \code{ExportToCellbrowser}:
+#' Write a big sparse matrix to a .tsv.gz file by writing chunks, concating them with the Unix cat command,
+#' then gziping the result. This does not work on Windows, we'd have to use the copy /b command there.
+#'
+#' @param inMat input matrix
+#' @param outFname output file name, has to end with .gz
+#' @param sliceSize=1000, size of each chunk in number of lines
+#'
+#' @return Invisibly returns \code{NULL}
+#'
+#' @importFrom data.table setDTthreads data.table fwrite
+#'
+#' @examples
+#' \dontrun{
+#' writeSparseMatrix(GetAssayData(pbmc_small, slot = "data"), "exprMatrix.tsv.gz")
+#' }
+#'
+writeSparseMatrix = function(inMat, outFname, sliceSize = 1000) {
+    # require(data.table)
+    fnames <- c()
+    setDTthreads(threads = 8)  # otherwise this would use dozens of CPUs on a fat server
+    mat = inMat
+    geneCount = nrow(mat)
+    message("Writing expression matrix to ", outFname)
+    startIdx = 1
+    while (startIdx < geneCount) {
+      endIdx <- min(startIdx + sliceSize - 1, geneCount)
+      matSlice <- mat[startIdx:endIdx, ]
+      denseSlice <- as.matrix(x = matSlice)
+      dt <- data.table(denseSlice)
+      dt <- cbind(gene = rownames(x = matSlice), dt)
+      writeHeader <- startIdx == 1
+      writeHeader <- FALSE
+      # if (startIdx == 1) {
+      #   writeHeader <- TRUE
+      # }
+      sliceFname <- paste0("temp", startIdx, ".txt")
+      fwrite(
+        dt,
+        sep = "\t",
+        file = sliceFname,
+        quote = FALSE,
+        col.names = writeHeader
+      )
+      fnames <- append(x = fnames, values = sliceFname)
+      startIdx <- startIdx + sliceSize
+    }
+    message("Concatenating chunks")
+    system(command = paste(
+      "cat",
+      paste(fnames, collapse = " "), "| gzip >",
+      outFname,
+      sep = " "
+    ))
```

Comment:
An OS check should be implemented in this function to prevent running on Windows

Reply:
done. did the check in the calling function.